File: loom-tools/src/loom_tools/common/config_resolver.py
"""Config resolution layer (#4039, Epic #3835 Phase 2).

A single resolver over the tier chain: private/shared defaults, the legacy
``.loom/config.json``, tracked ``.loom-project/project.json``, and ignored
``.loom-local/local.json``. See ``docs/design/config-resolution-tiers.md``
for the full design (schema, precedence table, and how this composes with
the existing **env > config > default** rule used by e.g. the ``autonomous``
block).

**This module is additive only.** No existing call site (e.g.
``loom_tools.common.paths._read_config_worktree_root``) is migrated onto it
in this PR. A follow-up issue migrates call sites one at a time.

Mirrors ``loom-daemon/src/config_resolver.rs`` and
``defaults/scripts/lib/config-resolver.sh`` key-for-key so the three
resolvers agree on one fixture tree (cross-language conformance test).

Example::

    from pathlib import Path
    from loom_tools.common.config_resolver import (
        get_path,
        resolve_effective_config,
    )

    effective = resolve_effective_config(Path("/path/to/repo"))
    enabled = get_path(effective, "autonomous.workFinder.enabled")
"""

from __future__ import annotations

import os
from pathlib import Path
from typing import Any

#: Repo-relative path to the legacy, currently load-bearing config file.
#: Every existing repo has (at most) this tier populated.
LEGACY_CONFIG_REL = ".loom/config.json"

#: Repo-relative path to the tracked, project-specific config file
#: introduced by Epic #3835. Empty (absent) in every repo until a repo
#: migrates onto the new layout (Phase 6).
PROJECT_CONFIG_REL = ".loom-project/project.json"

#: Repo-relative path to the ignored, host-local config override file.
#: Empty (absent) in every repo until Phase 3+.
LOCAL_CONFIG_REL = ".loom-local/local.json"

#: Env var overriding the private/shared defaults file location. Set to a
#: non-empty path to point at an alternate file; set to the empty string to
#: disable this tier entirely.
PRIVATE_DEFAULTS_ENV = "LOOM_CONFIG_DEFAULTS_FILE"

#: Home-relative default location of the private/shared defaults file. Epic
#: #3835 Phase 3 introduces the machine-level checkout this lives in; until
#: that phase ships the file simply does not exist on any host, so this
#: tier resolves to "contributes nothing" everywhere today.
_DEFAULT_PRIVATE_DEFAULTS_REL = ".local/share/loom/config/defaults.json"


def private_defaults_path() -> Path | None:
    """Resolve the path to the private/shared defaults file.

    Honors :data:`PRIVATE_DEFAULTS_ENV`. Returns ``None`` when the env var
    is explicitly set to an empty string (tier disabled) or when no home
    directory can be determined.

    Returns:
        The resolved path (which may or may not exist on disk), or
        ``None`` when this tier is disabled.
    """
    if PRIVATE_DEFAULTS_ENV in os.environ:
        override = os.environ[PRIVATE_DEFAULTS_ENV]
        return Path(override) if override else None
    try:
        home = Path.home()
    except RuntimeError:
        return None
    return home / _DEFAULT_PRIVATE_DEFAULTS_REL


def _soft_read_json_object(path: Path) -> dict[str, Any]:
    """Soft-fail JSON-object read.

    A missing file, an unreadable file, malformed JSON, or a non-object
    top-level value all resolve to an empty dict (``{}``) — that tier
    simply contributes nothing to the merge. Never a hard error, mirroring
    the soft-fail contract already used by
    ``loom_tools.common.paths._read_config_worktree_root``.

    Args:
        path: Path to the candidate JSON file.

    Returns:
        The parsed object, or ``{}`` on any failure.
    """
    # Imported lazily to avoid a module-level import cycle
    # (common.state has no dependency on config_resolver, but keeping the
    # import lazy here matches the existing convention in
    # common.paths._read_config_worktree_root).
    from loom_tools.common.state import read_json_file

    data = read_json_file(path, default={})
    return data if isinstance(data, dict) else {}


def deep_merge(base: Any, overlay: Any) -> Any:
    """Deep-merge ``overlay`` onto ``base``, returning the merged value.

    Semantics (matching ``jq``'s ``*`` operator exactly, so the Bash
    resolver can implement this via ``jq -n '$a * $b * $c * $d'`` and stay
    in lockstep):

    - When both sides are dicts, merge recursively key-by-key.
    - Otherwise, ``overlay`` replaces ``base`` outright — including when
      ``overlay`` is ``None``/``null`` (an explicit null in a higher tier
      means "clear this key", not "leave the lower tier's value").

    Args:
        base: The lower-precedence value.
        overlay: The higher-precedence value.

    Returns:
        The merged value. Never mutates ``base`` or ``overlay`` in place.
    """
    if isinstance(base, dict) and isinstance(overlay, dict):
        merged = dict(base)
        for key, value in overlay.items():
            if key in merged:
                merged[key] = deep_merge(merged[key], value)
            else:
                merged[key] = value
        return merged
    return overlay


def resolve_effective_config(repo_root: Path) -> dict[str, Any]:
    """Resolve the full effective config tree for ``repo_root``.

    Deep-merges, lowest to highest precedence:

    1. Private/shared defaults (:func:`private_defaults_path`) — a no-op
       tier on every host until Epic #3835 Phase 3 ships.
    2. Legacy ``.loom/config.json`` — the sole tier every existing repo has.
    3. Tracked ``.loom-project/project.json`` — new in #4039; empty until a
       repo migrates (Phase 6).
    4. Ignored ``.loom-local/local.json`` — new in #4039; host-local
       override.

    A missing or malformed file at **any** tier soft-fails to that tier
    contributing nothing, and this function never raises. When only tier 2
    has content — the state of every repo today — the result is exactly
    that file's parsed content: byte-for-byte identical to reading
    ``.loom/config.json`` directly, satisfying the #4039
    behavior-preservation requirement.

    Args:
        repo_root: The absolute repository root path.

    Returns:
        The merged effective config dict.
    """
    effective: dict[str, Any] = {}

    defaults_path = private_defaults_path()
    if defaults_path is not None:
        effective = deep_merge(effective, _soft_read_json_object(defaults_path))

    effective = deep_merge(effective, _soft_read_json_object(repo_root / LEGACY_CONFIG_REL))
    effective = deep_merge(effective, _soft_read_json_object(repo_root / PROJECT_CONFIG_REL))
    effective = deep_merge(effective, _soft_read_json_object(repo_root / LOCAL_CONFIG_REL))

    return effective


_MISSING = object()


def get_path(config: dict[str, Any], dotted: str, default: Any = None) -> Any:
    """Look up a dotted key path in an already-resolved effective config tree.

    Args:
        config: The merged effective config (from
            :func:`resolve_effective_config`).
        dotted: A dotted key path, e.g. ``"autonomous.workFinder.enabled"``.
        default: Value to return when the path is absent or indexes through
            a non-dict value.

    Returns:
        The resolved value, or ``default`` when the path can't be
        resolved.
    """
    cur: Any = config
    for segment in dotted.split("."):
        if not isinstance(cur, dict):
            return default
        cur = cur.get(segment, _MISSING)
        if cur is _MISSING:
            return default
    return cur

File: mcp-loom/src/shared/config-resolver.ts
/**
 * Config resolution layer for mcp-loom (Epic #3835, issue #4064).
 *
 * A TypeScript port of `loom-daemon/src/config_resolver.rs` (the reference
 * implementation). It resolves the effective config by deep-merging the tier
 * chain, lowest to highest precedence:
 *
 *   1. Private/shared defaults (`$LOOM_CONFIG_DEFAULTS_FILE`, default
 *      `~/.local/share/loom/config/defaults.json`) — a no-op tier on every
 *      host until Epic #3835 Phase 3 ships.
 *   2. Legacy `.loom/config.json` — the sole tier every existing repo has.
 *   3. Tracked `.loom-project/project.json` — empty until a repo migrates.
 *   4. Ignored `.loom-local/local.json` — host-local override.
 *
 * The tier order, `deepMerge` semantics (objects merge recursively, arrays
 * replace, explicit `null` replaces), and soft-skip behavior (missing /
 * unreadable / malformed / non-object files contribute nothing, never fatal)
 * are kept byte-compatible with the Rust reference. The shared conformance
 * fixture in `loom-tools/tests/fixtures/config_resolver/` must resolve
 * identically here and in Rust/Python/Bash.
 *
 * NOTE: all existing resolvers (Rust/Python/Bash) are read-only. This module
 * is likewise read-only — it never writes. The write-target policy for
 * `configure_terminal` is enforced separately via `hasNonLegacyTier` (see
 * issue #4064's write-policy decision): the writer refuses to flatten the
 * merged view back into the legacy file when a non-legacy tier is present.
 */

import { readFile } from "node:fs/promises";
import { homedir } from "node:os";
import { join } from "node:path";

/** Repo-relative path to the legacy, currently load-bearing config file. */
export const LEGACY_CONFIG_REL = ".loom/config.json";

/** Repo-relative path to the tracked, project-specific config file. */
export const PROJECT_CONFIG_REL = ".loom-project/project.json";

/** Repo-relative path to the ignored, host-local config override file. */
export const LOCAL_CONFIG_REL = ".loom-local/local.json";

/** Env var overriding the private/shared defaults file location. */
export const PRIVATE_DEFAULTS_ENV = "LOOM_CONFIG_DEFAULTS_FILE";

/** Home-relative default location of the private/shared defaults file. */
const DEFAULT_PRIVATE_DEFAULTS_REL = ".local/share/loom/config/defaults.json";

/** A JSON object (the shape every tier is normalized to before merging). */
export type JsonObject = Record<string, unknown>;

/**
 * True for a plain JSON object — i.e. not `null`, not an array, not a scalar.
 * Mirrors Rust's `Value::Object(_)` match arm.
 */
function isPlainObject(value: unknown): value is JsonObject {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

/**
 * Resolve the path to the private/shared defaults file, honoring
 * {@link PRIVATE_DEFAULTS_ENV}. Returns `null` when the env var is explicitly
 * set to an empty string (tier disabled). Mirrors Rust's
 * `private_defaults_path`: an unset env var falls back to the home-relative
 * default; a set-but-empty env var disables the tier.
 */
export function privateDefaultsPath(): string | null {
  const env = process.env[PRIVATE_DEFAULTS_ENV];
  if (env !== undefined) {
    return env === "" ? null : env;
  }
  return join(homedir(), DEFAULT_PRIVATE_DEFAULTS_REL);
}

/**
 * Soft-fail JSON-object read: a missing file, an unreadable file, malformed
 * JSON, or a non-object top-level value all resolve to an empty object
 * (`{}`) — that tier simply contributes nothing to the merge. Never throws.
 * Mirrors Rust's `soft_read_json_object`.
 */
export async function softReadJsonObject(path: string): Promise<JsonObject> {
  let text: string;
  try {
    text = await readFile(path, "utf-8");
  } catch {
    return {};
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(text);
  } catch {
    return {};
  }

  return isPlainObject(parsed) ? parsed : {};
}

/**
 * Deep-merge `overlay` onto `base`, returning the merged value.
 *
 * Semantics (identical to Rust's `deep_merge` and `jq`'s `*` operator):
 * - When both sides are JSON objects, merge recursively key-by-key.
 * - Otherwise `overlay` replaces `base` outright — including when `overlay`
 *   is `null` (an explicit `null` in a higher tier clears the key) and when
 *   either side is an array (arrays replace, never concatenate).
 */
export function deepMerge(base: unknown, overlay: unknown): unknown {
  if (isPlainObject(base) && isPlainObject(overlay)) {
    const merged: JsonObject = { ...base };
    for (const key of Object.keys(overlay)) {
      merged[key] = Object.prototype.hasOwnProperty.call(merged, key)
        ? deepMerge(merged[key], overlay[key])
        : overlay[key];
    }
    return merged;
  }
  return overlay;
}

/**
 * Resolve the full effective config tree for `repoRoot` by deep-merging the
 * tier chain lowest → highest. A missing or malformed file at any tier
 * soft-fails to that tier contributing nothing. When only the legacy tier has
 * content — the state of every repo today — the result is byte-for-byte the
 * parsed content of `.loom/config.json`, preserving existing behavior.
 */
export async function resolveEffectiveConfig(repoRoot: string): Promise<JsonObject> {
  let effective: JsonObject = {};

  const defaultsPath = privateDefaultsPath();
  if (defaultsPath !== null) {
    effective = deepMerge(effective, await softReadJsonObject(defaultsPath)) as JsonObject;
  }

  effective = deepMerge(effective, await softReadJsonObject(join(repoRoot, LEGACY_CONFIG_REL))) as JsonObject;
  effective = deepMerge(effective, await softReadJsonObject(join(repoRoot, PROJECT_CONFIG_REL))) as JsonObject;
  effective = deepMerge(effective, await softReadJsonObject(join(repoRoot, LOCAL_CONFIG_REL))) as JsonObject;

  return effective;
}

/**
 * Look up a dotted key path (e.g. `"autonomous.workFinder.enabled"`) in an
 * already-resolved effective config tree. Returns `undefined` on any missing
 * segment or when a non-object value is indexed further. Mirrors Rust's
 * `get_path`.
 */
export function getPath(config: unknown, dotted: string): unknown {
  let cur: unknown = config;
  for (const segment of dotted.split(".")) {
    if (!isPlainObject(cur) || !Object.prototype.hasOwnProperty.call(cur, segment)) {
      return undefined;
    }
    cur = cur[segment];
  }
  return cur;
}

/**
 * True when a non-legacy tier (private defaults, `.loom-project/project.json`,
 * or `.loom-local/local.json`) contributes any content for `repoRoot`.
 *
 * This is the guard behind the #4064 write-policy decision (option 3): when a
 * non-legacy tier is present, `configure_terminal` must NOT write the merged
 * view back to `.loom/config.json`, because doing so would flatten every
 * higher-tier override into the legacy file and de-sync it from its source of
 * truth. The writer refuses instead, directing the operator to edit the tier
 * file directly.
 */
export async function hasNonLegacyTier(repoRoot: string): Promise<boolean> {
  const defaultsPath = privateDefaultsPath();
  if (defaultsPath !== null) {
    if (Object.keys(await softReadJsonObject(defaultsPath)).length > 0) {
      return true;
    }
  }
  if (Object.keys(await softReadJsonObject(join(repoRoot, PROJECT_CONFIG_REL))).length > 0) {
    return true;
  }
  if (Object.keys(await softReadJsonObject(join(repoRoot, LOCAL_CONFIG_REL))).length > 0) {
    return true;
  }
  return false;
}
